### ISSUE TITLE
`gi-gtk-demo-layout` is disabled / out of date

### ISSUE THREAD

**Issue by deepfire** [haskell-gi/gi-gtk-examples#3]
After fixing the trivial symbol jitter issues, like this:
```
diff --git a/inputmethod/Layout.hs b/inputmethod/Layout.hs
index 788f435..0df4fa2 100644
--- a/inputmethod/Layout.hs
+++ b/inputmethod/Layout.hs
@@ -24,10 +24,12 @@ import GI.Pango
        (AttrList, Attribute, attrListInsert, attrListNew, Layout,
         layoutSetWidth, layoutNew, layoutSetAttributes, layoutSetText,
         layoutSetWrap)
-import GI.PangoCairo.Functions (showLayout, fontMapGetDefault)
+import GI.Gdk.Structs.Rectangle (getRectangleWidth)
+import GI.Gdk.Structs.EventKey (getEventKeyState, getEventKeyKeyval)
+import GI.PangoCairo.Functions (showLayout)
+import GI.PangoCairo.Interfaces.FontMap (fontMapGetDefault)
 import GI.Gdk
-       (keyvalToUnicode, keyvalName, eventKeyReadKeyval,
-        eventKeyReadState, EventKey, rectangleReadWidth)
+       (keyvalToUnicode, keyvalName, EventKey)
 import GI.Cairo.Structs.Context (Context(..))
 import Foreign.ForeignPtr (withForeignPtr)
 import Control.Monad.Trans.Reader (ReaderT(..))
@@ -190,8 +192,8 @@ interpretKeyPress e = do
     where returnJust = return . Just

 shiftAttribute :: Int -> Attribute -> IO Attribute
-shiftAttribute x attr = attr { paStart = x + paStart attr,
-                               paEnd   = x + paEnd attr }
+shiftAttribute x attr = attr -- { paStart = x + paStart attr,
+                             --   paEnd   = x + paEnd attr }
```
..the more complex issue of type ancestry arises:
```
[nix-shell:~/src/gi-gtk-examples]$ cabal build gi-gtk-demo-layout
Preprocessing executable 'gi-gtk-demo-layout' for gi-gtk-examples-0.15.0...
[1 of 1] Compiling Main             ( inputmethod/Layout.hs, dist/build/gi-gtk-demo-layout/gi-gtk-demo-layout-tmp/Main.o )

inputmethod/Layout.hs:101:10: error:
    • Required ancestor ‘GI.Pango.Objects.Context.Context’ not found for type ‘GI.Pango.Objects.FontMap.FontMap’.
    • In a stmt of a 'do' block: lay <- layoutNew ctxt
      In the expression:
        do { Gtk.init;
             win <- windowNew WindowTypeToplevel;
             onWidgetDestroy win mainQuit;
             area <- drawingAreaNew;
             .... }
      In an equation for ‘main’:
          main
            = do { Gtk.init;
                   win <- windowNew WindowTypeToplevel;
                   onWidgetDestroy win mainQuit;
                   .... }

inputmethod/Layout.hs:109:19: error:
    • Couldn't match expected type ‘IO a2’
                  with actual type ‘GHC.Int.Int32 -> m2 ()’
    • Probable cause: ‘layoutSetText’ is applied to too few arguments
      In a stmt of a 'do' block: layoutSetText lay (displayBuffer buffer)
      In the expression:
        do { layoutSetText lay (displayBuffer buffer);
             layoutSetAttributes lay [] }
      In a case alternative:
          Nothing
            -> do { layoutSetText lay (displayBuffer buffer);
                    layoutSetAttributes lay [] }

inputmethod/Layout.hs:110:43: error:
    • Couldn't match expected type ‘Maybe AttrList’
                  with actual type ‘[t1]’
    • In the second argument of ‘layoutSetAttributes’, namely ‘[]’
      In a stmt of a 'do' block: layoutSetAttributes lay []
      In the expression:
        do { layoutSetText lay (displayBuffer buffer);
             layoutSetAttributes lay [] }

inputmethod/Layout.hs:111:20: error:
    • Couldn't match type ‘GHC.Int.Int32’ with ‘Int’
      Expected type: (T.Text, [[Attribute]], GHC.Int.Int32)
        Actual type: (T.Text, [[Attribute]], Int)
    • In the pattern: (str, attrs, pos)
      In the pattern: Just (str, attrs, pos)
      In a case alternative:
          Just (str, attrs, pos)
            -> do { layoutSetText lay (displayBufferPreedit buffer str pos);
                    attrListNewFromList
                      (map (shiftAttribute (cursor + 1)) (concat attrs))
                    >>= layoutSetAttributes lay }

inputmethod/Layout.hs:112:19: error:
    • Couldn't match expected type ‘IO a3’
                  with actual type ‘GHC.Int.Int32 -> m3 ()’
    • Probable cause: ‘layoutSetText’ is applied to too few arguments
      In a stmt of a 'do' block:
        layoutSetText lay (displayBufferPreedit buffer str pos)
      In the expression:
        do { layoutSetText lay (displayBufferPreedit buffer str pos);
             attrListNewFromList
               (map (shiftAttribute (cursor + 1)) (concat attrs))
             >>= layoutSetAttributes lay }
      In a case alternative:
          Just (str, attrs, pos)
            -> do { layoutSetText lay (displayBufferPreedit buffer str pos);
                    attrListNewFromList
                      (map (shiftAttribute (cursor + 1)) (concat attrs))
                    >>= layoutSetAttributes lay }

inputmethod/Layout.hs:113:40: error:
    • Couldn't match type ‘IO Attribute’ with ‘Attribute’
      Expected type: [Attribute]
        Actual type: [IO Attribute]
    • In the first argument of ‘attrListNewFromList’, namely
        ‘(map (shiftAttribute (cursor + 1)) (concat attrs))’
      In the first argument of ‘(>>=)’, namely
        ‘attrListNewFromList
           (map (shiftAttribute (cursor + 1)) (concat attrs))’
      In a stmt of a 'do' block:
        attrListNewFromList
          (map (shiftAttribute (cursor + 1)) (concat attrs))
        >>= layoutSetAttributes lay

inputmethod/Layout.hs:114:68: error:
    • Couldn't match type ‘AttrList’ with ‘Maybe AttrList’
      Expected type: AttrList -> IO ()
        Actual type: Maybe AttrList -> IO ()
    • In the second argument of ‘(>>=)’, namely
        ‘layoutSetAttributes lay’
      In a stmt of a 'do' block:
        attrListNewFromList
          (map (shiftAttribute (cursor + 1)) (concat attrs))
        >>= layoutSetAttributes lay
      In the expression:
        do { layoutSetText lay (displayBufferPreedit buffer str pos);
             attrListNewFromList
               (map (shiftAttribute (cursor + 1)) (concat attrs))
             >>= layoutSetAttributes lay }

inputmethod/Layout.hs:121:10: error:
    Variable not in scope:
      rectangleReadWidth
        :: GI.Gdk.Structs.Rectangle.Rectangle -> IO Integer

inputmethod/Layout.hs:171:5: error:
    • Couldn't match expected type ‘Render ()’
                  with actual type ‘a0 -> m0 ()’
    • Probable cause: ‘showLayout’ is applied to too few arguments
      In a stmt of a 'do' block: showLayout lay
      In the expression:
        do { moveTo 0 0;
             showLayout lay }
      In an equation for ‘updateArea’:
          updateArea area lay
            = do { moveTo 0 0;
                   showLayout lay }

inputmethod/Layout.hs:171:16: error:
    • Couldn't match expected type ‘Context’ with actual type ‘Layout’
    • In the first argument of ‘showLayout’, namely ‘lay’
      In a stmt of a 'do' block: showLayout lay
      In the expression:
        do { moveTo 0 0;
             showLayout lay }

inputmethod/Layout.hs:175:18: error:
    Variable not in scope: eventKeyReadState :: EventKey -> IO [t0]

inputmethod/Layout.hs:177:19: error:
    Variable not in scope: eventKeyReadKeyval :: IO GHC.Word.Word32

inputmethod/Layout.hs:179:31: error:
    • Couldn't match type ‘GHC.Word.Word32’ with ‘Int’
      Expected type: IO Int
        Actual type: IO GHC.Word.Word32
    • In the second argument of ‘(<$>)’, namely
        ‘keyvalToUnicode keyVal’
      In a stmt of a 'do' block:
        keyChar <- toEnum <$> keyvalToUnicode keyVal
      In the expression:
        do { keyVal <- eventKeyReadKeyval;
             keyName <- keyvalName keyVal;
             keyChar <- toEnum <$> keyvalToUnicode keyVal;
             case keyChar of {
               '\NUL'
                 -> case keyName of {
                      "Left" -> ...
                      "Right" -> ...
                      "BackSpace" -> ...
                      _ -> ... }
               ch -> do { ... } } }

inputmethod/Layout.hs:195:25: error:
    • Couldn't match expected type ‘IO Attribute’
                  with actual type ‘Attribute’
    • In the expression: attr
      In an equation for ‘shiftAttribute’: shiftAttribute x attr = attr
```

**Comment by deepfire** [comment]
cc @hamishmack

**Comment by deepfire** [comment]
Versions:
```
[nix-shell:~/src/gi-gtk-examples]$ git log -n1 --oneline
ddf999b Use haskell-gi 0.20 and add travis files

[nix-shell:~/src/gi-gtk-examples]$ ghc-pkg list | grep gi
    gi-atk-2.0.11
    gi-cairo-1.0.11
    gi-gdk-3.0.11
    gi-gdkpixbuf-2.0.11
    gi-gio-2.0.11
    gi-glib-2.0.11
    gi-gobject-2.0.11
    gi-gtk-3.0.11
    gi-gtk-hs-0.3.4.0
    gi-pango-1.0.11
    gi-pangocairo-1.0.11
    haskell-gi-0.20
    haskell-gi-base-0.20
```

**Comment by hamishmack** [comment]
I have done the easy bits.  We might need @garetxe to look at the hard part (which is where I got stuck before).  We need to be able convert from AttrList to [Attribute] and back again.  I think we will need an AttrList iterator and I am not sure if [these functions](https://developer.gnome.org/pango/stable/pango-Text-Attributes.html#pango-attr-list-get-iterator) are included in the GObject introspection data.

**Comment by deepfire** [comment]
The chief reason I have looked at this example is because it was the only one that demonstrated the use of PangoCairo.  Maybe the attribute shuffling part is something of a secondary value, maybe not.

**Comment by garetxe** [comment]
I took a look,  efb9bfd  now compiles (with `haskell-gi` and `haskell-gi-base` from git HEAD). Not sure what it is supposed to be doing, so no idea if it's working :)

@hamishmack `pango_attr_list_get_iterator` is marked as not introspectable, so we do not generate bindings for it. We could add an override and generate bindings, but I'd rather not: the API is simply unsafe to use from any garbage-collected language (I guess that's why it is marked as not introspectable):
```haskell
iter <- attrListGetIterator list
-- XXX GC runs here, discarding `list`
attrs <- attrIteratorGetAttrs iter
```
is unsafe, since `list` may be garbage collected between the two lines, and then `iter` is invalid.

So in the demo I opted to write a safe wrapper by hand, perhaps we could move it to `gi-gtk-hs`.
